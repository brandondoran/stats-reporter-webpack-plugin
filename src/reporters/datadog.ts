import axios from 'axios';
import * as gzipSize from 'gzip-size';
import * as path from 'path';
import { StatsReporter } from '../types';

const API_URL = 'https://app.datadoghq.com/api/v1/series';
const METRIC_TYPE_GAUGE = 'gauge';

const headers = { 'Content-Type': 'application/json' };
const method = 'POST';
const getTimestamp = () => parseInt(`${new Date().getTime() / 1000}`, 10);

export interface DataDogStatsReporterOptions {
  apiKey: string;
  gzipSize: boolean;
  metricName: string;
  tags?: string[];
  test?: RegExp;
}

export class DataDogStatsReporter implements StatsReporter {
  private readonly apiKey: string;
  private readonly gzipSize: boolean;
  private readonly metricName: string;
  private readonly tags: string[];
  private readonly test?: RegExp;
  private readonly url: string;

  constructor(options: DataDogStatsReporterOptions) {
    this.apiKey = options.apiKey;
    this.gzipSize = options.gzipSize;
    this.metricName = options.metricName;
    this.tags = options.tags || [];
    this.test = options.test;
    this.url = `${API_URL}?api_key=${this.apiKey}`;

    this.validateOptions();
  }

  public async send(stats: any) {
    try {
      const series = await this.parseStats(stats);
      // console.log(require('util').inspect(series, { showHidden: false, depth: null }));
      return await axios({
        data: { series },
        headers,
        method,
        url: this.url
      });
    } catch (err) {
      throw new Error(`DataDogStatsReporter: ${err.message}`);
    }
  }

  private filterAssets(stats: any): string[] {
    const { assets } = stats;
    if (!this.test) {
      return assets;
    }
    return assets.filter((asset: any) => this.test!.test(asset.name));
  }

  private parseStats(stats: any) {
    const { outputPath } = stats;
    const assets = this.filterAssets(stats);
    const now = getTimestamp();
    const promises = assets.map(async (asset: any) => {
      const size =
        this.gzipSize === false
          ? asset.size
          : await gzipSize.file(path.join(outputPath, asset.name));
      return {
        metric: `${this.metricName}.bytes${path.extname(asset.name)}`,
        points: [[now, size]],
        tags: [
          ...this.tags,
          `chunk:${asset.chunkNames[0]}`,
          `type:${path.extname(asset.name).substring(1)}`
        ],
        type: METRIC_TYPE_GAUGE
      };
    });
    return Promise.all(promises);
  }

  private validateOptions(): void {
    const errors = [];
    if (!this.apiKey) {
      errors.push('apiKey is required.');
    }
    if (!this.metricName) {
      errors.push('metricName is required.');
    }
    if (!Array.isArray(this.tags)) {
      errors.push('tags must be an array of strings.');
    }
    if (errors.length !== 0) {
      throw new Error(`DataDogStatsReporter: ${errors.join(' ')}`);
    }
  }
}
