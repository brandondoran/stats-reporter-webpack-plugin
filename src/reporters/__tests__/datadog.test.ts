import axios from 'axios';
import { DataDogStatsReporter, DataDogStatsReporterOptions } from '../datadog';

jest.mock('axios');

describe('DataDogStatsReporter', () => {
  let reporter: DataDogStatsReporter;
  let options: DataDogStatsReporterOptions;

  beforeEach(() => {
    options = {
      apiKey: 'test-key',
      gzipSize: true,
      metricName: 'testapp.assets'
    };
    reporter = new DataDogStatsReporter(options);
  });

  describe('constructor', () => {
    it('should be an instance of DataDogStatsReporter', () => {
      expect(reporter).toBeInstanceOf(DataDogStatsReporter);
    });

    it('should throw if options.apiKey is missing', () => {
      delete options.apiKey;
      expect(
        () => new DataDogStatsReporter(options)
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if options.metricName is missing', () => {
      delete options.metricName;
      expect(
        () => new DataDogStatsReporter(options)
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw if multiple options are missing', () => {
      delete options.apiKey;
      delete options.metricName;
      expect(
        () => new DataDogStatsReporter(options)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
