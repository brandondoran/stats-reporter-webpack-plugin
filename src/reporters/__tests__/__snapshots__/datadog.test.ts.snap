// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataDogStatsReporter constructor should throw if multiple options are missing 1`] = `"DataDogStatsReporter: apiKey is required. metricName is required."`;

exports[`DataDogStatsReporter constructor should throw if options.apiKey is missing 1`] = `"DataDogStatsReporter: apiKey is required."`;

exports[`DataDogStatsReporter constructor should throw if options.metricName is missing 1`] = `"DataDogStatsReporter: metricName is required."`;
